presto-main/src/main/java/com/facebook/presto/operator/HashBuilderOperator.java

```diff
@@ -538,35 +543,58 @@ private void unspillLookupSourceIfRequested()
         }
 
         verify(spiller.isPresent());
-        verify(!unspillInProgress.isPresent());
-
-        localUserMemoryContext.setBytes(getSpiller().getSpilledPagesInMemorySize() + index.getEstimatedSize().toBytes(), enforceBroadcastMemoryLimit);
-        unspillInProgress = Optional.of(getSpiller().getAllSpilledPages());
+        if (isSpillInSmallMemoryPool(operatorContext.getSession())) {
+            localUserMemoryContext.setBytes(index.getEstimatedSize().toBytes(), enforceBroadcastMemoryLimit);
+        }
+        else {
+            verify(!unspillInProgress.isPresent());
+            localUserMemoryContext.setBytes(getSpiller().getSpilledPagesInMemorySize() + index.getEstimatedSize().toBytes(), enforceBroadcastMemoryLimit);
+            unspillInProgress = Optional.of(getSpiller().getAllSpilledPages());
+        }
 
         state = State.INPUT_UNSPILLING;
     }
 
     private void finishLookupSourceUnspilling()
     {
         checkState(state == State.INPUT_UNSPILLING);
-        if (!unspillInProgress.get().isDone()) {
-            // Pages have not be unspilled yet.
-            return;
+        if (isSpillInSmallMemoryPool(operatorContext.getSession())) {
```

Comment:
what's the downside of this change?  why do we only do it for spillInsmallMemoryPool

Reply:
Eventually I am planning to remove `INPUT_UNSPILLING` and related code completely, but I prefer to do it gradually just in case there is regression. Another reason is to make testing and rollout easier, I can test and experiment with multiple changes without worrying about breaking release.